ships/0024-surfacing-errors-to-buildrun.md

```diff
@@ -0,0 +1,186 @@
+
+SPDX-License-Identifier: Apache-2.0
+-->
+
+---
+title: surfacing-errors-to-buildrun
+authors:
+  - "@dalbar"
+reviewers:
+  - "@qu1queee"
+  - "@HeavyWombat"
+approvers:
+  - "@adambkaplan"
+  - "@sbose78"
+creation-date: 2021-09-10
+last-updated: 2021-09-10
+status: draft
+
+---
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [docs](/docs/)
+
+## Summary
+
+This ship provides a mechanism to surface errors in the build process
+into BuildRun objects granting more visibility into
+without being Tekton/Implementation specific. To
+achieve this goal, we introduce two new generic results
+`shp-error-reason` and `shp-error-message`, that a TaskRun can emit
+and a BuildRun can pick up. 
+
+
+## Motivation
+A strategy's container logs differ for its steps and utilized
+tooling. Therefore, finding an error message and extracting the error
+reason is challenging for machine processing and consumption from
+third parties due to the logs heterogeneous nature and the need to parse
+`stdout`. 
+For a better user experience, we propose that a strategy author
+can define the failure reason in a compact way that can
+be represented in the BuildRun status, for example, through TaskRun results.
+Build strategy authors are already aware of this tooling as they can emit the size
+digest of the output image similarly, see [the BuildKit strategy](https://github.com/shipwright-io/build/blob/main/samples/buildstrategy/buildkit/buildstrategy_buildkit_cr.yaml#L65).
+
+Additionally, the proposal enables a user to keep the troubleshooting
+process on the BuildRun level and only descend further into Tekton
+specifics when necessary.
+
+### Goals
+
+- Allow build strategy users to define error details that will show up on the BuildRun status.
+- Allow our own steps like the git step to also use this mechanism to provide additional error details in the BuildRun status
+
+### Non-Goals
+
+- The specifics of errors regarding tooling are not part of this
+  proposal. Neither is their means for reporting them. Errors highly
+  depend on each strategy
+
+## Proposal
+
+We propose to introduce two generic results for errors:
+`shp-error-message` and `shp-error-reason`.  The result
+`shp-error-reason` has a machine-recognizable value that is ideally prefixed
+with the problematic technology, for example, `git` in case of missing
+a secret. The result `shp-error-message` should be descriptive text to
+give the user enough information to troubleshoot the BuildRun.  A task
+run emits the new results as Tekton results.  
+Finally, we must extend the BuildRun controller
+logic to propagate the new results into the BuildRun CRD's
+`status.failure` property.
+
+BuildRun with new error results: 
+
+```yaml
+status:
+    failure:
```

Comment:
So I wonder about surfacing yet another top level item in our BuildRun status.

1) It is already somewhat expansive.  Consider https://github.com/shipwright-io/build/blob/fc3975da9fafb1f1525c8856ada78a4e5ab17c6b/pkg/apis/build/v1alpha1/buildrun_types.go#L114-L151

2) the aforementioned Conditions field at https://github.com/shipwright-io/build/blob/fc3975da9fafb1f1525c8856ada78a4e5ab17c6b/pkg/apis/build/v1alpha1/buildrun_types.go#L127 is an array of a type that already has status/reason/message at https://github.com/shipwright-io/build/blob/ac139dba9437bca165137856e18a593417945630/pkg/apis/build/v1alpha1/buildrun_types.go#L241-L261

3) we already have top level failure semantics at https://github.com/shipwright-io/build/blob/fc3975da9fafb1f1525c8856ada78a4e5ab17c6b/pkg/apis/build/v1alpha1/buildrun_types.go#L150-L157 ... I point to the field in status, and then that rangee ^^ includes the definition of the struct `FailedAt`

I think a new user coming in and seeing out BuildRun status have a top level field called `failure` and a peer pointer field called `failedAt` is going to think that is sloppy.  

Minimally, given we have a pointer `failedAt` field, since failures are not a given, the reason/message should be coupled with that existing field some how.

And yes, I see the prototype has it as a pointer as well:  https://github.com/shipwright-io/build/pull/901/files#diff-993203175e9d10b489c9c9dc6e9d54966c4864fb2f6b79a126bcee8cf0dda3d0R150

But I still see no need why we need 2 pointers to provide all the relative metadata around failures

I *suspect* maybe the additive / 2 pointers was done to be less of disruptive API change.  If I am correct in that suspicion, while I appreciate the desire for that, we are still at alpha, and I don't want to sacrifice potential API sprawl for that.

Reply:
Thanks for your input. It seems like I should work on the alternatives section of the ship to capture our thoughts about the right place to surface errors.

Variant 1:

I agree that the status field is already expanded a lot. The second point with two pointers is also a clear disadvantage of this variant. 
The only advantage compared to the third variant is that the existing API is preserved which seems to be a non-factor since we are in alpha. Also, adding two more fields to `failedAt` isn't really disruptive.

Variant 2:

Seems to be the most idiomatic approach in K8s world. Other controllers can also consume them without worrying about our CRD. Tools like `kubectl` can also use its existing features for conditions to react to failures e.g. `wait`. Based on the conventions for conditions I see some possible downsides.

From [community repo](https://github.com/kubernetes/community/blob/4f95b9210536ee62165e0a186ef5fab19a8be63f/contributors/devel/sig-architecture/api-conventions.md):

> Conditions should complement more detailed information about the observed status of an object written by a controller, rather than replace it

+1 for more details on why it failed

> Controllers should apply their conditions to a resource the first time they visit the resource, even if the status is Unknown. This allows other components in the system to know that the condition exists and the controller is making progress on reconciling that resource.

Conditions are something that is always present in the status. I don't think it is too noisy, but it may be a concern for others. During build it would be `Unknown` and for successful build it would change to `False`.

> Condition type names should describe the current observed state of the resource, rather than describing the current state transitions. This typically means that the name should be an adjective ("Ready", "OutOfDisk") or a past-tense verb ("Succeeded", "Failed") rather than a present-tense verb ("Deploying"). Intermediate states may be indicated by setting the status of the condition to Unknown.

Good naming will be key to avoid confusion. Additionally, conditions and their name become a fixed part of the API. I want to think about a possible condition to capture our use-case. If the above points can be addressed, then I would prefer variant 2.

Variant 3) 

As expressed in variant 1, I think it makes sense to combine failure related information in one field. So for now, I would eliminate the first variant.


### In summary 

I will investigate variant two and try to come up with a condition. The ship will be updated and the `alternatives` sections will be expanded.

Reply:
I am still uncertain about the `type`. I am open for suggestions (my initial thoughts: ImageBuildSucceded/StrategySucceded/ExecutionSucceded/BuildStrategySucceded). 

A possible set of conditions for a failed build in yaml:

```yaml
conditions:
  - lastTransitionTime: "2021-10-12T16:57:07Z"
    message: |
      "step-build-and-push" exited with code 1 (image: "icr.io/obs/codeengine/buildkit/builder@sha256:a11e2348f9ee40822fc28dcb501c57cd02ebd31fb441841bfe5c144cc9d77fc6"); for logs run: kubectl -n baran-test logs apparmor-build-buildrun-gkxld-zrrtv-pod-8qpkt -c step-build-and-push
    reason: Failed
    status: "False"
    type: Succeeded
  - lastTransitionTime: "2021-10-12T16:57:07Z"
    message: |
      The key is invalid for the specified target. Please make sure that the remote source exists, you have sufficient rights and the key is in the right format.
    reason: GitAuthSSH
    status: "False"
    type: ImageBuildSucceded
```